src/qibolab/instruments/icarusqfpga.py

```diff
@@ -1,259 +1,453 @@
+        self,
+        qubits: Dict[QubitId, Qubit],
+        couplers,
+        sequence: PulseSequence,
+        options: ExecutionParameters,
+    ):
+        """Plays the pulse sequence on the IcarusQ RFSoC and awaits acquisition
+        at the end.
 
         Arguments:
-            waveform (numpy.ndarray): Numpy array of size DAC_CHANNELS X DAC_SAMPLE_SIZE with type signed short.
+            qubits (dict): Dictionary of qubit IDs mapped to qubit objects.
+            sequence (PulseSequence): Pulse sequence to be played on this instrument.
+            options (ExecutionParameters): Object representing acquisition type and number of shots.
         """
+        super().play(qubits, couplers, sequence, options)
+        self.device.set_adc_trigger_repetition_rate(int(options.relaxation_time / 1e3))
+        readout_pulses = list(
+            filter(lambda pulse: pulse.type is PulseType.READOUT, sequence.pulses)
+        )
+        readout_qubits = {pulse.qubit for pulse in readout_pulses}
+
+        if options.acquisition_type is AcquisitionType.RAW:
+            self.device.set_adc_trigger_mode(0)
+            self.device.arm_adc(self.adcs_to_read, options.nshots)
+            raw = self.device.result()
+            return self.process_readout_signal(raw, readout_pulses, qubits, options)
+
+        # Currently qunit only supports single qubit readout demodulation
+        elif options.acquisition_type is AcquisitionType.INTEGRATION:
+            self.device.set_adc_trigger_mode(1)
+            self.device.set_qunit_mode(0)
+            raw = self.device.start_qunit_acquisition(options.nshots, readout_qubits)
+
+            if options.averaging_mode is not AveragingMode.SINGLESHOT:
+                res = {
+                    qubit: IntegratedResults(I + 1j * Q).average
+                    for qubit, (I, Q) in raw.items()
+                }
+            else:
+                res = {
+                    qubit: IntegratedResults(I + 1j * Q)
+                    for qubit, (I, Q) in raw.items()
+                }
+
+            for ro_pulse in readout_pulses:
+                res[ro_pulse.serial] = res[ro_pulse.qubit]
+            return res
+
+        elif options.acquisition_type is AcquisitionType.DISCRIMINATION:
+            self.device.set_adc_trigger_mode(1)
+            self.device.set_qunit_mode(1)
+            raw = self.device.start_qunit_acquisition(options.nshots, readout_qubits)
+            res = {qubit: SampleResults(states) for qubit, states in raw.items()}
+            for ro_pulse in readout_pulses:
+                res[ro_pulse.serial] = res[ro_pulse.qubit]
+            return res
+
+    def play_sequences(
+        self,
+        qubits: Dict[QubitId, Qubit],
+        couplers,
+        sequences: List[PulseSequence],
+        options: ExecutionParameters,
+    ):
+        """Wrapper for play."""
+        return [
+            self.play(qubits, couplers, sequence, options) for sequence in sequences
+        ]
```

Comment:
I would keep this in user's code, or rely on `execute_pulse_sequences`.

Since it's not part of the interface, this would not go through the platform, and it would require explicit invocation, directly accessing the instrument.

Reply:
This was done for compatibility with an older version of the compatibility class.

Resolved in c08af5d